### ISSUE TITLE
Updating grid function coefficients every time step

### ISSUE THREAD

**Issue by JeremiahDBrown** [mfem/mfem#3156]
I am trying to implement perfectly matched layer boundary conditions in a time-domain simulation (see Jin, "The Finite Element Method in Electromagnetics").  I am trying to use the overall PML concept within a E/B update framework similar to the maxwell miniapp.  After some derivations, there are extra terms in the time-dependent update equation (time derivative of E), whose weak form is:
$-\sum_j {\int_\Omega W_i\cdot\overline{K}_2\cdot\psi_j(t)\cdot W_j d\Omega}$
where
$\psi_j (t+\Delta t) = (1 - e^{-\frac{\bar{\sigma}\Delta t}{\epsilon}}) e_j(t) + e^{-\frac{\bar{\sigma}\Delta t}{\epsilon}} \psi_j(t)$
Both $\overline{K}_2$ and $\bar{\sigma}$ are time-independent diagonal matrices that have nonzero values in a boundary region surrounding the main simulation region.  That boundary region is part of the overall mesh but is flagged so that we can use restricted coefficients to limit the updates to those regions.  The terms other than $\psi_j(t)$ can be implemented using a VectorMassIntegrator.  My concern is how to efficiently implement the update equation for $\psi_j(t)$ though.  It is presumably represented as a gridfunction, as is $e_j(t)$.  Given that I am using ND and RT elements, it seems that I may have to wrap both gridfunctions in a vectorgridfunctioncoefficient and then do a new projection every time step.

Is there a better approach?

**Comment by mlstowell** [comment]
Hi, @JeremiahDBrown,

It seems to me that the update for $\psi$ is linear and $\psi$ is in H(Curl) just as $E$ is.  If this is the case then this could be carried out as an operation on the `GridFunctions` themselves. In other words just treat the `GridFunctions` $\psi$ and $e$ as vectors and compute the linear combination:
```
psi *= exp(-sigma * t / epsilon);
psi.Add(1.0 - exp(-sigma * t / epsilon), e);
```

Best wishes,
Mark

**Comment by JeremiahDBrown** [comment]
@mlstowell ,
Thanks for looking at this.  I realized I accidentally replaced $\Delta t$ with $t$ in the exponential.  Just edited that.  It doesn't change anything significant, but just making sure things are correct.
Unfortunately though, sigma in the exponential is also a vector coefficient.  Jin's approach is have 3 separate vectors, psi_x, psi_y, and psi_z so that he can multiply by a scalar quantity (exponentials based on sigma_x, sigma_y or sigma_z), but I don't think that works with MFEM's ND or RT elements (scalar coefficients do not project onto gridfunctions when these elements are used).

**Comment by mlstowell** [comment]
Hi, @JeremiahDBrown,

Yes, I see $\sigma$ does complicate things...

Splitting into the vector components will also complicate matters. If you can ensure that the mesh in the PML region has one set of edges perpendicular to the outer boundary and the other two parallel to it you may be able to simplify things. This would involve determining which ND dofs correspond to edges perpendicular to the boundary and treating these differently as you loop through the vectors and update $\psi$ entry by entry. My though is that if you can distinguish these two sets of edges you can use different values for $\sigma$ as needed.

One of my colleagues, @psocratis, has more experience with PMLs than I do so let me ping him. He may see another option.

Best wishes,
Mark

**Comment by psocratis** [comment]
@stefanhenneking, I think you had to deal with something like this recently, didn't you?

**Comment by stefanhenneking** [comment]
Yes, I had to implement something similar but was prototyping it in fenics, not mfem. However, in my case I was looking at a time-dependent PML for acoustics with two keys differences to this case: the extra unknown $\psi$ was scalar-valued and the PML only acting one-directional (i.e. $\sigma$ only acting as $\sigma_x$) in which case @mlstowell's suggestion would work. So I'm not sure what the optimal way is to compute the PML equation most efficiently here but would also be interested in the best solution.

**Comment by JeremiahDBrown** [comment]
For the time being, I've implemented this as outlined above.  $\psi$ is a grid function, and I've created a ConvolutionCoefficient derived from VectorCoefficient.  The Eval method uses the GetVectorVal method to obtain values of $\psi$ and $e$ to compute the value to be projected.  Every time step, I tell the coefficient to make a backup copy of $\psi$, as once I call ProjectCoefficient, the results of psi->GetVectorValue are no longer valid.

This doesn't seem very efficient, but I can't come up with a better approach right now.  It seems to be working as expected.
